New plots

javierggt opened the pull request:
## Description

This improves the plots:
- new plots
- plots include uncertainties
- The code to produce the plots is split into smaller functions
- the calculation of quantiles, p_acq, etc is decoupled from the plotting functions
- added a config class to hold global options

## Interface impacts
None

## Testing
<!-- If relevant describe any special setup for testing. -->

### Unit tests
<!-- At least one of these must be checked if unit tests exist. DELETE the unchecked/untested options. -->
- [x] No unit tests
- [ ] Mac
- [ ] Linux
- [ ] Windows

Independent check of unit tests by [REVIEWER NAME]
- [ ] [PLATFORM]:

### Functional tests

In a dedicated environment, I installed chandra_aca master (required for binomial uncertainty). There, I created and empty `$SKA/data/acq_stat_reports` and copied `/proj/sot/ska/data/acq_stat_reports/acq_fail_fitfile.json` into it (useless but still needed). Then I ran:
```
acq-stat-reports -v debug --url /aspect/test_review_outputs/acq_stat_reports/pr-6/web
acq-stat-reports-toc
acq-stat-reports-summary
```
[output](https://icxc.cfa.harvard.edu/aspect/test_review_outputs/acq_stat_reports/pr-6/www/2024/YEAR/index.html)

jeanconn commented:
I see you've made some tests - is the idea these would just be run as part of package development only?  Or for CI add to ska_testr and install this package so it gets tested?

javierggt commented:
No strong opinion, but I would not add it to ska_testr. If they become standard, and are used in other places, then we have the tests ready. Meanwhile, we can use them if we make changes to the function.

jeanconn commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,532 @@
+    return np.percentile(n, quantiles)
+
+
+def get_histogram_quantile_ranges(  # noqa: PLR0915
```
This seems like it belongs someplace else (ska_helpers? ).

javierggt replied:
that would be the idea, to move it somewhere else, but right now it just works for this package. Getting it to a state where it works for other packages still requires work.

Maybe moving it into `ska_trend.utils` makes sense as a utility with a restricted scope (for trending).

jeanconn replied:
Sounds good to let this go in this PR then and keep brainstorming about how to optimize the placement of the trending codes .

javierggt replied:
Yes, the thing with this is one can think of little options to add to the function, and the options might be different for different packages.

First we need to know exactly which options are general, and how to best apply to special cases before trying to generalize.

jeanconn commented:
With regard to the tests, it looks like there needs to be an `__init__.py` in acq_stat_reports/tests for the pytest magic to work.  And at that point, the tests are running but aren't passing for me.  Let me know if you @javierggt  need more debug info on that or if just running the tests also fails for you.

jeanconn commented on acq_stat_reports/acq_stat_reports.py:
```diff
@@ -134,184 +111,211 @@ def make_acq_plots(acqs, tstart=0, tstop=None, outdir=None, close_figures=False)
+    figscale=(2, 1),
+)
+def fail_rate_plot(data, **kwargs):  # noqa: ARG001 (kwargs is neeeded by the decorator)
+    import warnings
```
Isn't this already imported?

javierggt replied:
Yes. It was.

jeanconn commented on acq_stat_reports/acq_stat_reports.py:
```diff
@@ -5,8 +5,10 @@
 
+import agasc
 import jinja2
 import matplotlib.pyplot as plt
```
Does it also make sense to use Tom's new plot-script convention of 
```
# Force non-interactive Agg backend for matplotlib in the context of this script
os.environ["MPLBACKEND"] = "Agg"
```
?

javierggt replied:
Where is that used?

The problem is that this needs to be set before matplotlib is imported. There was a time when `matplotlib.use('agg')` had to be called before importing matplotlib, but I just checked and it works even after I already created another figure.

I think that means the correct way is to use `matplotlib.use('agg')` in `main()`, not at module level.

jeanconn replied:
@taldcroft suggested it to me for https://github.com/sot/ska_trend/blob/26441a05574e8f593db5b51fa93c9fdb80c1a9fe/ska_trend/wrong_box_anom/wrong_box_anom.py#L15 with the idea that for these codes that are meant to make plots etc and aren't meant to be imported that it would be simplest to do via the env var.

javierggt replied:
I just tried using `matplotlib.use('agg')` in main. That does not prevent the usual `RuntimeWarning: More than 20 figures have been opened.` I think it is good to add it. This would prevent the script from popping an undesired window.

As for the `close_figures` option/argument, this has an effect only when running inside ipython. If you call the function from ipython, it briefly opens a window and then closes it.

javierggt commented:
I fixed the broken test. I had not updated it after adding the 2-sigma quantiles. That is an argument in favor of automated testing in the PR.

javierggt commented on acq_stat_reports/templates/acq_stats/index.html:
```diff
@@ -48,10 +48,11 @@ <H3>{{ rep.human_date_start }} through {{ rep.human_date_stop }}</H3>
 <TABLE>
-<TR><TD><img src="mag_histogram.png"></TD><TD><img src="zoom_mag_histogram.png"></TD></TR>
-<TR><TD><img src="mag_pointhist.png"></TD><TD><img src="zoom_mag_pointhist.png"></TD></TR>
+<TR><TD><img src="mag_pointhist.png"></TD><TD><img src="t_ccd_pointhist.png"></TD></TR>
```
Right now I didn't need to.

jeanconn commented on acq_stat_reports/acq_stat_reports.py:
```diff
@@ -134,184 +111,209 @@ def make_acq_plots(acqs, tstart=0, tstop=None, outdir=None, close_figures=False)
+    plot_params = {
+        "acq_stars": {
+            "data": "two_year_acqs",
+            "class": "acq_stars_plot",
```
Since these are no longer classes, maybe it is worth the time to relabel them as funcs or whatever.

jeanconn commented:
If these are benign I think they still need to get handled (but probably in chandra_aca?) 
```
(ska3-flight-latest) flame:data jean$ acq-stat-reports
2024-09-26 14:28:58,799 main: ---------- acq stat reports update at 2024-09-26 18:28:58.799 ----------
/Users/jean/miniforge3/envs/ska3-flight-latest/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:696: RuntimeWarning: invalid value encountered in _beta_isf
  return _boost._beta_isf(x, a, b)
/Users/jean/miniforge3/envs/ska3-flight-latest/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:696: RuntimeWarning: invalid value encountered in _beta_isf
  return _boost._beta_isf(x, a, b)
/Users/jean/miniforge3/envs/ska3-flight-latest/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:696: RuntimeWarning: invalid value encountered in _beta_isf
  return _boost._beta_isf(x, a, b)
/Users/jean/miniforge3/envs/ska3-flight-latest/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:696: RuntimeWarning: invalid value encountered in _beta_isf
  return _boost._beta_isf(x, a, b)
/Users/jean/miniforge3/envs/ska3-flight-latest/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:696: RuntimeWarning: invalid value encountered in _beta_isf
  return _boost._beta_isf(x, a, b)
/Users/jean/miniforge3/envs/ska3-flight-latest/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:696: RuntimeWarning: invalid value encountered in _beta_isf
  return _boost._beta_isf(x, a, b)
```

javierggt commented:
> If these are benign I think they still need to get handled (but probably in chandra_aca?)
> 
> ```
> (ska3-flight-latest) flame:data jean$ acq-stat-reports
> 2024-09-26 14:28:58,799 main: ---------- acq stat reports update at 2024-09-26 18:28:58.799 ----------
> /Users/jean/miniforge3/envs/ska3-flight-latest/lib/python3.11/site-packages/scipy/stats/_continuous_distns.py:696: RuntimeWarning: invalid value encountered in _beta_isf
> ```

Yes, I don't see those, but they should be handled in chandra_aca

jeanconn commented:
Did @taldcroft put that minimal example somewhere?  I'm surprised you aren't seeing these but maybe they are only on apple silicon.

taldcroft commented:
See https://github.com/sot/chandra_aca/issues/176 for some clarity on the RuntimeWarning.

taldcroft commented on acq_stat_reports/config.py:
```diff
@@ -0,0 +1,21 @@
+class Config:
+    # "Top-level data directory",
+    data_dir = "."
```
Maybe `web_dir_default` or `plot_dir_default`? I think it is currently being used in that way.

javierggt replied:
Same comment as above. I agree that this needs changing, but I already started #9 and I don't want to start making changes that cause conflict with the PR downstream.

taldcroft commented on acq_stat_reports/acq_stat_reports.py:
```diff
@@ -38,18 +43,29 @@ def get_parser():
+        type=int,
+    )
     parser.add_argument(
         "-v",
```
Out of scope for this PR, but modern code should use `--log-level`. At some point we should change this.

javierggt replied:
I added this change to #9

taldcroft reviewed (changes requested):
Overall this is a great improvement. Here are some comments / questions / suggestions. I haven't fully reviewed all of it, but did hit most of the high points.

javierggt commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+    """
+    if quantiles is None:
+        quantiles = [15.9, 50, 84.1]
+    samples = (
+        np.random.uniform(size=n_realizations * len(p_acq_model)).reshape(
+            (-1, len(p_acq_model))
+        )
+        < p_acq_model[None]
+    )
```
I will fix this. Basically, you have N probabilities `p_acq`, and you want M draws with a binomial probability `p_acq`. That should be a (M, N) array. `np.random.uniform` here generates a 1d array with shape. It is better like this:
```python
samples = (
        np.random.uniform(size=(len(p_acq_model), n_realizations) < p_acq_model[np.newaxis]
    )

```

javierggt replied:
I guess this conversation can be resolved.

javierggt commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+    prob_column="p_acq_model",
+):
+    """
+    Summarize data in bins and calculate quantiles.
```
This function handles exactly one dataset, so there might be a misunderstanding.

The looping over columns is to produce N-dimensional histograms. In other words, if you give bins with two columns, this function will not produce the (unrelated) columns corresponding to 1d histograms. It will produce a single 2d histogram.

Take a look at the test.

javierggt commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+def get_histogram_quantile_ranges(  # noqa: PLR0915
+    data,
+    bin_edges,
+    extra_cols=(),
```
Losely speaking, this function requires three basic ingredients, as you said in another comment: `successes`, `probs` and `bins`, but sometimes you want to get the mean, min and max for other quantities that are not directly related to the bins.

A typical example is plotting the mean `p_acq` for each magnitude bin. `p_acq` is not in the bin definition, so it does not appear in the result, but if you want to make that plot, you add it to `extra_cols`.

I have used it these days, it just didn't end up in a final plot.

javierggt replied:
one convention could be to get the max, min and mean of all columns in the table, so the extra_cols argument is not needed. The the function calls need to be changed to only include the columns that we want in the result.

taldcroft replied:
> A typical example is plotting the mean p_acq for each magnitude bin. p_acq.

I think this would be best done with a call to `scipy.stats.binned_statistic`. The goal is to make functions that are minimal and do just one thing. 

But having now spent a huge amount of time staring at this function, I can at least see what `extra_cols` is doing. So if that is your preference then that's fine, but it needs to be documented carefully because the outputs are different.

javierggt replied:
I made changed in [0a012a](https://github.com/sot/acq_stat_reports/pull/6/commits/0a012a7dfa0a2d34c23239efdc63bdfe7b1a1cb9) to make this function more readable. I'm not sure how `scipy.stats.binned_statistic` will make this significantly different from the latest version.

javierggt commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+    extra_cols = list(extra_cols)
+    bin_cols = [f"{bin_col}_bin" for bin_col in cols]
+    for bin_col in cols:
+        bin_edges[bin_col] = np.atleast_1d(bin_edges[bin_col])
```
Strictly speaking, a single value would have only the underflow and overflow bin, and nothing would fail.

javierggt commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+            raise ValueError("bin_edges must be a dict of 1D arrays")
+
+    # data = data.copy()
+    bin_edges = bin_edges.copy()
```
This is the part related to my answer above (the apparent misunderstanding).

This does an n-dimensional binning. The code is hard to understand, and it was at this point when I decided I really needed to add a test. I think it can be improved, but it will not necessarily be "simpler".

javierggt replied:
> It looks like this application only uses 1-d binning (am I missing something?). That shaped my understanding of what this function should be doing and the docs did not change my prior.

I have made 2d histograms, to make plots as a function if T_ccd and magnitude. These plots have not been good enough, so they never made it to the final reports. I don't intend to remove this functionality.

The code is not easy to read. I agree. It might be better using an off-the-shelf function, be it `binned_statistic_dd` or something else. I don't think I would use `binned_statistic_dd` though.

Again: I do not intend to remove the n-dimensional binning functionality. We did not write formal requirements before I wrote it, but it is one of the requirements in my view.

taldcroft replied:
Great, I'm looking forward to seeing an updated version that uses OTS code for the heavy lifting, hopefully from a widely-used package already in Ska.

javierggt commented on acq_stat_reports/config.py:
```diff
@@ -0,0 +1,21 @@
+    close_figures = False
+
+    # "Do not include bad stars in the reports",
+    remove_bad_stars = False
```
bad stars will make things look worse than they are. To me, the real question is why should we include bad stars?

taldcroft replied:
The question of removing bad stars depends critically on the question you are asking, or in the case of this app, which plot you are making. 

If the question is "How accurate is the acq. prob model assuming we know the star magnitudes?" then we want to remove bad stars. That's what I do in all the model calibration and validation work. Eventually I'd like to see a plot in the acq_stats_report that looks like one of the validation plots that we look at for model promotion, with `remove_bad_stars=True`.

If the question is "In flight statistics, are we acquiring as many stars as we expect from the model?" then we should not remove bad stars. That is because encountering bad stars is an unavoidable fact of flight operations. If 50% of stars were bad then that would have a big impact on the answer to this question and we need to know that.

So that would imply that removing bad stars is an attribute of each plot (where a plot answers a question). The most common will be `remove_bad_stars=False`, but I think it would be best to be explicit about this and not define a default.

javierggt replied:
I removed this option in PR #9

javierggt commented on acq_stat_reports/config.py:
```diff
@@ -0,0 +1,21 @@
+    figure_width = 5
+
+    # "Default figure height in inches",
+    figure_height = 2.5
```
Specifying `figure_width` and `aspect_ratio` or `figure_width` and `figure_height` is a matter of choice.

The `scale` allows you to do anything arbitrary, so I do not understand the point of it not being square. If you look at the figures we have in the page, the ones that take two columns are not twice as high as the single-column ones.

javierggt replied:
ah, and this is not "for each plot" this is a default that applies to every plot in order to keep it consistent.

taldcroft replied:
We agree that only 2 numbers are required to specify the size of a plot, but your code uses 4 independent numbers:  figure_width, figure_height (from the configuration), scale_width, scale_height (hard coded as `figscale`). Why does the code need 4 numbers when 2 are sufficient?

javierggt replied:
I think this can be resolved. Can it?

taldcroft commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+
+    g = data.group_by(bin_cols)
+    assert len(g.groups.indices) <= len(quantiles) + 1  # sanity check
+    idx = g[["bin"] + cols].groups.aggregate(np.mean)["bin"].astype(int)
```
Why is this aggregating over `cols` but then ignoring the result by using only the `bin` column?

javierggt replied:
The way the algorithm works is by assigning a unique bin ID, so when you group by `cols`, each group should have a unique bin ID, which is stored in the `bin` column.

taldcroft replied:
FWIW, this passes tests if I change the line to:
```
idx = g[["bin"]].groups.aggregate(np.mean)["bin"].astype(int)
# Or even simpler
idx = g["bin"].groups.aggregate(lambda c: c[0])
# Or even faster
idx = g["bin"][g.groups.indices[:-1]]
```

javierggt commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+    following columns:
+    - `bin`: the unique bin index
+    - `n`: the number of samples in the bin
+    - `{col}_low`: the lower edge of the bin (repeated for all bin_edges keys)
+    - `{col}_high`: the upper edge of the bin (repeated for all bin_edges keys)
+    - `{col}`: the center of the bin (repeated for all bin_edges keys)
+    - `{col}_delta`: the width of the bin (repeated for all bin_edges keys)
```
The bin size is not necessarily the same in all bins. That is why it is not in meta. For example, the plot of acq success fraction versus magnitude in the acq stat report uses variable binning.

javierggt replied:
I think this can be resolved.

taldcroft commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+from acq_stat_reports.config import conf
+
+
+class BinnedData:
```
Seems like a nice opportunity to use a dataclass, but at the very least this needs a docstring.

javierggt replied:
I changed it so it is a dataclass, but I'm not sure it adds much in this case.

taldcroft commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+    def __init__(self, data, bins, extra_cols=None, selection=None):
+        self.bins = bins
+        self.selection = selection if selection is not None else (lambda x: x)
+        self.data = self.selection(data)
```
It's not intuitive that the `data` attribute gets auto-filtered.
```
bd = BinnedData(data, bins, selection=...)
# unexpected that bd.data is different from data.
```

javierggt replied:
By default it was not auto-filtered. It was auto-filtered if you gave it the option, in which case it would be pretty much what you expect.

Anyway, I removed the selection in [85ea8a](https://github.com/sot/acq_stat_reports/pull/6/commits/85ea8a41578d21eb98a34e16ce9db513699b6885)

javierggt commented on acq_stat_reports/acq_stat_reports.py:
```diff
@@ -134,184 +111,209 @@ def make_acq_plots(acqs, tstart=0, tstop=None, outdir=None, close_figures=False)
+        ),
+    }
+
+    functions = {
```
I made these changes in [15ef4e](https://github.com/sot/acq_stat_reports/pull/6/commits/15ef4e560a5f760145c8e5390facf1ca0d4fbf50)

javierggt commented on acq_stat_reports/acq_stat_reports.py:
```diff
@@ -134,184 +111,209 @@ def make_acq_plots(acqs, tstart=0, tstop=None, outdir=None, close_figures=False)
+    }
+
+    plot_params = {
+        "acq_stars": {
```
I made these changes in [15ef4e](https://github.com/sot/acq_stat_reports/pull/6/commits/15ef4e560a5f760145c8e5390facf1ca0d4fbf50)

javierggt commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+    plt.xlim(0, 1)
+
+
+def _get_quantiles_(p_acq_model, quantiles=None, n_realizations=10000):
```
Made this change in [c6566a](https://github.com/sot/acq_stat_reports/pull/6/commits/c6566a57a1c07ca6d32d0d800047319138632508)

javierggt commented on acq_stat_reports/config.py:
```diff
@@ -0,0 +1,21 @@
+    figure_height = 2.5
+
+    # "Close matplotlib figures after plotting",
+    close_figures = False
```
Remember that this PR is about getting new plots. I wanted to be done with this more or less quickly so I could improve the plots themselves and perhaps use plotly, at which point this will not matter.

Still, I added it to PR #9

javierggt commented on acq_stat_reports/config.py:
```diff
@@ -0,0 +1,21 @@
+class Config:
```
One of the things I removed from the class when you started criticizing it was the `__repr__` method, which I have used to give feedback to stdout when running the script.

javierggt commented on acq_stat_reports/utils.py:
```diff
@@ -0,0 +1,538 @@
+
+def mpl_plot(**defaults):
+    """
+    Decorator to handle common matplotlib plotting tasks.
```
Did this in [8f0074](https://github.com/sot/acq_stat_reports/pull/6/commits/8f007483cb643473e902a9904a362d54f540afb0)

javierggt commented on acq_stat_reports/acq_stat_reports.py:
```diff
@@ -535,12 +582,12 @@ def main():
             "prev": f"{args.url}/{prev_range['year']}/{prev_range['subid']}/index.html",
         }
 
+        conf.data_dir = str(webout)
```
Remember that this PR is for adding plots. I did change some other things, but that does not mean they will not change anymore.

In particular, I think there should not be two output directories. There should be only one, and it should be called `output_dir`. The split between "data" and "web" is not necessary, because all output from this package should be suitable for a web page, and that includes json files.

Anyway, I made some changes in this direction in #9 but there will be more changes, and I don't want to make these changes in this PR.

javierggt commented on acq_stat_reports/acq_stat_reports.py:
```diff
@@ -471,17 +473,62 @@ def get_data():
     return all_acq
 
 
 def main():
```
I added this in PR #9

javierggt commented:
@taldcroft I think I addressed your comments. Some of the changes went into PR #9 because in here they would cause conflicts, and they had nothing to do with plots. A few others would not cause conflict, but still they were closer to the changes in PR #9.

taldcroft reviewed (approved):
OK, good enough. I still have some comments but I'll move over to #9.
